### PR TITLE
feat(pt): DPLR in PyTorch backend

### LINKED ISSUE #3490: [Feature Request] pt: support DPLR in the PyTorch backend

### LINKED ISSUE #5121: [Feature Request] DPLR in PyTorch backend
### Summary

Dear developers, 

I have developed [DPLR in PyTorch backend](https://github.com/ChiahsinChu/torch-dplr) in the plugin mode. This implementation has passed tests and been ready to use. The reason I have not started a PR  to the main repository is that:
1. This implementation depends on two PRs (#4661 and #4628). 
2. This implementation depends on a third-party package [torch-admp](https://github.com/ChiahsinChu/torch-admp) for PME calculation. 

I would like to know your suggestions about this feature, keeping it as a stand-alone plugin or merging it to the main repository.

Best regards.

### Detailed Description

See above.

### Further Information, Files, and Links

_No response_

### PR DESCRIPTION
**Pull request by ChiahsinChu**
This PR introduces the DPLR (Dipole Charge) modifier implementation for the PyTorch backend of DeepMD-kit, enabling quantum mechanical charge distribution modeling in molecular dynamics simulations.

## Key Changes
- **New DipoleChargeModifier class** (`deepmd/pt/modifier/dipole_charge.py`):
  - Implements the dipole charge modifier for PyTorch backend with autograd

- **Dependency addition**:
  - Added `torch_admp` package to project dependencies in `pyproject.toml`
  - Enables advanced molecular dynamics with polarizable force fields

- **Comprehensive testing** (`source/tests/pt/modifier/test_dipole_charge.py`):
  - Tests for JIT compilation compatibility
  - Consistency checks between TensorFlow and PyTorch implementations
  - Serialization/deserialization validation
  - Training workflow verification

## Benefits
- Maintains compatibility with existing TensorFlow implementation
- Supports periodic boundary conditions with proper Ewald summation
- Provides a unified interface across different backends

## Files Modified
- 8 files changed, 577 insertions(+), 1 deletion(-)
- Added comprehensive test suite with water molecule test cases
- Updated package dependencies and module initialization

## Summary by CodeRabbit

* **New Features**
  * DipoleChargeModifier for dipole–charge corrections
  * DPModifier to support DP-based model modifiers

* **Serialization / Inference**
  * Modifier serialize/deserialize support; inference loading updated to use serialized modifier data

* **Documentation**
  * DPLR docs now list PyTorch alongside TensorFlow and note limitations

* **Tests**
  * Comprehensive tests for dipole modifier, serialization round-trips, cross-backend consistency, and training

* **Chores**
  * Updated PyTorch packaging pins and inference entrypoint handling

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Review by njzjz — COMMENTED** [review]
Need documentation.

File: source/tests/pt/modifier/test_dipole_charge.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            box=t_box,
+        )
+        e, f, v = self.dm_tf.eval(
+            coord=coord,
+            box=box,
+            atype=atype.reshape(-1),
+        )
+
+        np.testing.assert_allclose(
+            to_numpy_array(dm_pred["energy"]).reshape(-1), e.reshape(-1), rtol=1e-4
```
Comment (wanghan-iapcm) [inline source/tests/pt/modifier/test_dipole_charge.py]: Should we tighten the tolerance for these values? Given that we are using float64, an absolute tolerance of $10^{-6}$ should be achievable. 

While the current relative tolerance is set to $10^{-4}$, I suspect the presence of small force components may be the limiting factor preventing further reduction.

File: source/tests/pt/modifier/test_dipole_charge.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        self.sys_charge_map = [6.0, 1.0]
+        self.descriptor_dict = {
+            "type": "se_e2_a",
+            "sel": [46, 92],
+            "rcut_smth": 0.5,
+            "rcut": 4.00,
+            "neuron": [
+                25,
+                50,
+            ],
```
Comment (wanghan-iapcm) [inline source/tests/pt/modifier/test_dipole_charge.py]: you may reduce the `sel` and `neuron` to accelerate the test.

File: deepmd/pt/modifier/dipole_charge.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+            Extended coordinates with shape (nframes, (natoms + nsel) * 3)
+            where nsel is the number of selected atoms
+        """
+        mask = make_mask(self.sel_type, atype)
+
+        nframes = coord.shape[0]
+        natoms = coord.shape[1] // 3
+
+        all_dipole = []
+        for ii in range(nframes):
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]: We should consider implementing batch inference rather than using a for loop to improve efficiency.

Note that we either:
1. impose that the atom types are consistent across all frames in the batch; 
2. or extend coordinates with the virtual atoms for un-selected atom, while set their charges to 0. 

approach 2 provides higher flexibility and maintainability to the code at potentially larger computational cost. 

Personally i would prefer approach 2

File: deepmd/pt/modifier/dipole_charge.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+            Dictionary containing the correction terms:
+            - energy: Energy correction tensor with shape (nframes, 1)
+            - force: Force correction tensor with shape (nframes, natoms, 3)
+            - virial: Virial correction tensor with shape (nframes, 3, 3)
+        """
+        if box is None:
+            raise RuntimeError(
+                "dipole_charge data modifier can only be applied for periodic systems."
+            )
+        else:
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]: We can simplify this logic by removing the `else` keyword, since the early raise makes it redundant.

File: deepmd/pt/modifier/dipole_charge.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+                input_coord,
+                atype,
+                input_box,
+                fparam,
+                aparam,
+            )
+
+            tot_e = []
+            # add Ewald reciprocal correction
+            for ii in range(nframes):
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]: Batch Ewald calculation is preferable to the for loop.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/pt/modifier/dipole_charge.py (hunk 4 above)
Comment (ChiahsinChu) [reply deepmd/pt/modifier/dipole_charge.py]: In the current implementation (pt backend), periodic is not required in the level of BaseModifier. I therefore need to add the raise here.

File: deepmd/pt/modifier/dipole_charge.py [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        Returns
+        -------
+        dict
+            The serialized data
+        """
+        data = {
+            "@class": "Modifier",
+            "type": self.modifier_type,
+            "@version": 3,
+            "model_name": self.model_name,
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]: model data should be serialized, otherwise the deserialization depends on  model files that may not be available.

File: deepmd/pt/modifier/dipole_charge.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+    """
+
+    def __new__(
+        cls, *args: tuple, model_name: str | None = None, **kwargs: dict
+    ) -> "DipoleChargeModifier":
+        return super().__new__(cls, model_name)
+
+    def __init__(
+        self,
+        model_name: str,
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]: input a model object may be a better choice.

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/pt/modifier/dipole_charge.py [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+        )
+        self.placeholder_pairs = torch.ones((1, 2), device=env.DEVICE, dtype=torch.long)
+        self.placeholder_ds = torch.ones((1), device=env.DEVICE, dtype=torch.float64)
+        self.placeholder_buffer_scales = torch.zeros(
+            (1), device=env.DEVICE, dtype=torch.float64
+        )
+
+        self.ewald_batch_size = ewald_batch_size
+        if dp_batch_size is None:
+            dp_batch_size = int(os.environ.get("DP_INFER_BATCH_SIZE", 1))
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]: The default value for batch size is for inference, while your implementation is supposed to be used for either training or inference.

File: deepmd/pt/modifier/dipole_charge.py [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+
+        Returns
+        -------
+        dict
+            The serialized data
+        """
+        dd = BaseModifier.serialize(self)
+        dd.update(
+            {
+                "model": self._model.serialize(),
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]:
```suggestion
                "dipole_model": self._model.serialize(),
```

File: deepmd/pt/modifier/dipole_charge.py [hunk 10]
```diff
@@ -0,0 +1,429 @@
...
+    def __init__(
+        self,
+        model_name: str | None,
+        model_charge_map: list[float],
+        sys_charge_map: list[float],
+        ewald_h: float = 1.0,
+        ewald_beta: float = 1.0,
+        ewald_batch_size: int = 5,
+        dp_batch_size: int | None = None,
+        model: DipoleModel | None = None,
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]:
```suggestion
        dipole_model: DipoleModel | None,
        dipole_model_file_name: str | None = None,
        model_charge_map: list[float],
        sys_charge_map: list[float],
        ewald_h: float = 1.0,
        ewald_beta: float = 1.0,
        ewald_batch_size: int = 5,
        dp_batch_size: int | None = None,
```

by default, user provides a model object, otherwise try to load a model from a file.

File: deepmd/pt/modifier/dipole_charge.py [hunk 11]
```diff
@@ -0,0 +1,429 @@
...
+        )
+        return dd
+
+    @classmethod
+    def deserialize(cls, data: dict) -> "DipoleChargeModifier":
+        data = data.copy()
+        data.pop("@class", None)
+        data.pop("type", None)
+        data.pop("@version", None)
+        model_obj = DipoleModel.deserialize(data.pop("model"))
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]:
```suggestion
        model_obj = DipoleModel.deserialize(data.pop("dipole_model"))
```

**Review by wanghan-iapcm — CHANGES_REQUESTED** [review]

File: deepmd/pt/infer/deep_eval.py [hunk 12]
```diff
@@ -172,19 +175,20 @@ def __init__(
...
+            if len(extra_files["modifier_data"]) > 0:
+                modifier_data = extra_files["modifier_data"]
                 if isinstance(modifier_data, bytes):
-                    modifier_data = io.BytesIO(modifier_data)
+                    modifier_data = pickle.loads(modifier_data)
                 # Load the modifier directly from the file-like object
-                modifier = torch.jit.load(modifier_data, map_location=env.DEVICE)
+                modifier = BaseModifier.get_class_by_type(
+                    modifier_data["type"]
+                ).deserialize(modifier_data)
```
[the remark this answers is not in the document]
Comment (ChiahsinChu) [reply deepmd/pt/infer/deep_eval.py:191]: The original dict, including np.ndarray, is dumped by `pickle.dumps`. I therefore need to load it with `pickle.loads`.

File: deepmd/pt/modifier/dipole_charge.py [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+    model_name
+            The model file for the DeepDipole model
+    model_charge_map
+            Gives the amount of charge for the wfcc
+    sys_charge_map
+            Gives the amount of charge for the real atoms
+    ewald_h
+            Grid spacing of the reciprocal part of Ewald sum. Unit: A
+    ewald_beta
+            Splitting parameter of the Ewald sum. Unit: A^{-1}
```
Comment (wanghan-iapcm) [inline deepmd/pt/modifier/dipole_charge.py]: please update the help message according to the updated interface.

**Review by wanghan-iapcm — APPROVED** [review]

File: pyproject.toml [hunk 14]
```diff
@@ -140,6 +140,9 @@ jax = [
     # The pinning of ml_dtypes may conflict with TF
     # 'jax-ai-stack;python_version>="3.10"',
 ]
+torch = [
+    "torch-admp",
```
Comment (njzjz) [inline pyproject.toml]: `>=1.1.5` should be also added here if it is required
